hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/flow/txn/logic/SchedulePurger.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.workflows.handle.flow.txn.logic;
+
+import com.hedera.node.app.service.schedule.ScheduleService;
+import com.hedera.node.app.service.schedule.WritableScheduleStore;
+import com.hedera.node.app.spi.metrics.StoreMetricsService;
+import com.hedera.node.app.workflows.dispatcher.WritableStoreFactory;
+import com.hedera.node.app.workflows.handle.ScheduleExpirationHook;
+import com.hedera.node.app.workflows.handle.flow.txn.UserTransactionComponent;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.Instant;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * The logic for the schedule service cron that purges expired schedules which runs as a part of default
+ * handle process for the user transaction
+ */
+@Singleton
+public class SchedulePurger {
+    private final ScheduleExpirationHook scheduleExpirationHook;
+    private final StoreMetricsService storeMetricsService;
+
+    /**
+     * Constructs the SchedulePurger with the given dependencies.
+     * @param scheduleExpirationHook the schedule expiration hook
+     * @param storeMetricsService the store metrics service
+     */
+    @Inject
+    public SchedulePurger(
+            @NonNull final ScheduleExpirationHook scheduleExpirationHook,
+            @NonNull final StoreMetricsService storeMetricsService) {
+        this.scheduleExpirationHook = scheduleExpirationHook;
+        this.storeMetricsService = storeMetricsService;
+    }
+
+    /**
+     * Expire schedules that are due to be executed between the last handled transaction time and the current consensus
+     * time.
+     * @param userTxnContext the user transaction component
+     */
+    public void expireSchedules(@NonNull UserTransactionComponent userTxnContext) {
+        final var lastHandledTxnTime = userTxnContext.lastHandledConsensusTime();
+        if (lastHandledTxnTime == Instant.EPOCH) {
+            return;
+        }
+        if (userTxnContext.consensusNow().getEpochSecond() > lastHandledTxnTime.getEpochSecond()) {
+            final var firstSecondToExpire = lastHandledTxnTime.getEpochSecond();
```

Comment:
These calculations look different than the original code. Does it calculate the same values?

Reply:
The original code was looking at first consensus time of last block and current time. But we can accomplish the same by using the last handled time and current time. So, we purge expired schedules for each transaction.